### ISSUE TITLE
Validator on field defined with Union of arbitrary type and with default value raises ConfigError

### ISSUE THREAD

**Issue by vbrinnel** [samuelcolvin/pydantic#1161]
# Question

Hi, 
I wonder if the behavior below results from a bug or me missing a point. 
A clarification would be appreciated.
Thanks

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
pydantic version: 1.3
pydantic compiled: False
install path: ~/miniconda3/lib/python3.8/site-packages/pydantic
python version: 3.8.0 | packaged by conda-forge | (default, Nov 22 2019, 19:11:19)  [Clang 9.0.0 (tags/RELEASE_900/final)]
platform: macOS-10.13.6-x86_64-i386-64bit
optional deps. installed: []
```

Customize pydantic BaseConfig and define arbitrary class _Test_:
```py
from pydantic import BaseModel, validator, Field, BaseConfig
from typing import Union, Type

BaseConfig.arbitrary_types_allowed = True 

class Test:
	pass
```

Defining class _A_ without validator works: 
```
class A(BaseModel):
	view: Union[Type[Test], str] = Test
```

Defining class _A_ with a validator fails:
```
class A(BaseModel):

	view: Union[Type[Test], str] = Test

	@validator("view")
	def validate(cls, v):
		return v
```

With the following error:
```
---------------------------------------------------------------------------
ConfigError                               Traceback (most recent call last)
<ipython-input-2-72c9cde31275> in <module>
----> 1 class A(BaseModel):
      2 
      3         view: Union[Type[Test], str] = Test
      4 
      5         @validator("view")

~/miniconda3/lib/python3.8/site-packages/pydantic/main.py in __new__(mcs, name, bases, namespace, **kwargs)
    221         if _custom_root_type:
    222             validate_custom_root_type(fields)
--> 223         vg.check_for_unused()
    224         if config.json_encoders:
    225             json_encoder = partial(custom_pydantic_encoder, config.json_encoders)

~/miniconda3/lib/python3.8/site-packages/pydantic/class_validators.py in check_for_unused(self)
    171         if unused_validators:
    172             fn = ', '.join(unused_validators)
--> 173             raise ConfigError(
    174                 f"Validators defined with incorrect fields: {fn} "  # noqa: Q000
    175                 f"(use check_fields=False if you're inheriting from the model and intended this)"

ConfigError: Validators defined with incorrect fields: validate (use check_fields=False if you're inheriting from the model and intended this)
```

Setting the default value to _Field(Test)_ avoids the ConfigError:
```
class A(BaseModel):

	view: Union[Type[Test], str] = Field(Test)

	@validator("view")
	def validate(cls, v):
		return v
```

**Comment by samuelcolvin** [comment]
I agree it's unclear, the problem is 

https://github.com/samuelcolvin/pydantic/blob/dbc044e3570170e30b1046474ac7040b1b100b07/pydantic/main.py#L141

Attributes of these types are ignored, thus the `view` field doesn't exist and the validator is correctly complaining that it's being asked to reference a field that doesn't exist. 

This check is because in the following example, we wouldn't expect `Foo` to be a field of the model:

```py
class MyModel(BaseModel):
    a: int
    b: str

    class Foo:
        pass
```

Obviously `thing = Foo` looks very different to `class Foo:...`, but they both come through as types aka `isintance(Foo, type) is True`.

Options to resolve this:
1. use `Field` as you have above which works fine
2. skip the `UNTOUCHED_TYPES` check on annotations but not attributes. That would fix your case, but would mean `thing: Type[Foo] = Foo` worked, but `thing = Foo` would silently fail, which might be even more confusing
3. only exclude subtypes of `type` if the name of the attribute matches the name of the type. This should exclude sub-classes while probably not effecting assignments like yours.
4. Keep the logic same but allow some escape hatch, e.g. a function on `Config` which replaces `isinstance(value, untouched_types)`

I'm inclined to to lend towards 3 and 4. Thoughts?

**Comment by vbrinnel** [comment]
Thanks for the clarification.
For the sake of completeness, please let me add that this also fails:

```
class A(BaseModel):

    view = Test

    @validator("view")
    def validate(cls, v):
        return v
```

which 3. will not be able to fix, correct ?

As for a potential 5.: requiring inner classes to feature a custom decorator (say _@innerclass_) is not something you would consider, would you ?

**Comment by samuelcolvin** [comment]
No, 3. should still fix that because the attribute is called "view" not "Test"

**Comment by PrettyWood** [comment]
Closed by #2214
